### PR TITLE
Use `getVisitor` in `ChangeDependency` over `getRecipeList` for reuse

### PR DESCRIPTION
**Pull request by timtebeek**
## What's your motivation?
Want to use this in `RelocatedDependencyCheck`, but even in this form it's not yet working.

## Any additional context
- https://github.com/openrewrite/rewrite-java-dependencies/pull/63#issuecomment-1902681113

### CODE REVIEW

File: src/main/java/org/openrewrite/java/dependencies/RelocatedDependencyCheck.java [hunk 1]
```diff
@@ -130,181 +128,177 @@ public TreeVisitor<?, ExecutionContext> getVisitor(Accumulator acc) {
...
-                        tag = maybeUpdate(acc, tag, groupId, artifactId, ctx);
+
+                    private <T extends Tree> T maybeUpdate(T tree, String groupId, @Nullable String artifactId, ExecutionContext ctx) {
+                        Relocation relocation = getRelocation(groupId, artifactId);
+                        if (relocation != null) {
+                            insertRow(groupId, artifactId, relocation, ctx);
+                            if (Boolean.TRUE.equals(changeDependencies) && artifactId != null) {
+                                String newGroupId = relocation.getTo().getGroupId();
+                                String newArtifactId = Optional.ofNullable(relocation.getTo().getArtifactId()).orElse(artifactId);
+                                doAfterVisit(new ChangeDependency(
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/dependencies/RelocatedDependencyCheck.java:265]: This `doAfterVisit` is the reason we can't for now have a shared `maybeUpdate` between the Maven and Gradle visitors. 🤔

**Comment by timtebeek** [comment]
Turns out merging the `maybeUpdate` methods isn't quite possible (or I've missed how up to now), but at least they each use the same `ChangeDependency` recipe now. That makes this a smaller iteration (apart from whitespace) but ready to review & merge.

**Comment by shanman190** [comment]
Do you remember what was happening when extracting the `maybeUpdate`? Curory review shows them to be exact duplicate code blocks, so lifting the method up to the top level where it would be visible to both and using it that way should result in identical behavior.

**Comment by timtebeek** [comment]
They are identical, but the `doAfterVisit` is called on each individual visitor. When we pull the method up then `doAfterVisit` is called on the TreeVisitor that calls the Maven and Gradle visitors, and the changes are not applied. 🤔 Passing the Maven/Gradle visitor in as argument to the method also doesn't work, as `doAfterVisit` is protected. While one _could_ create an interface that make it public, I think that might lead to more questions.

**Comment by shanman190** [comment]
Ahh, I see, I see. It's scoping of the protected method that then becomes the issue. What if we just forced the visit inline directly rather than using `doAfterVisit` at all?

**Comment by timtebeek** [comment]
Had considered that as well, but `maybeUpdate` only has a partial subtree, and the Gradle ChangeDependency visits the compilationUnit to set the gradle project field, and needs that to access for instance repositories to resolve dependencies.
Even for Maven the dependency doesn't get added when you merely visit the dependency tag, so we're near out of (likeable) options I think.

**Review by shanman190 — APPROVED** [review]

**Comment by timtebeek** [comment]
Thanks for helping think through the options! If I've missed any let me know.

**Pull request merged**